doc/developer/design/20240625_source_versioning__table_from_sources.md

```diff
@@ -379,7 +374,46 @@ for now. In the future we need to refactor the source dataflows to track progres
 on a per-export basis and do independent reclocking such that we can manage the
 timely capabilities / progress of each source export independently.
 
-#### Migration of source statements and collections
+Envelope and encoding/format options are currently specified on a per-source level.
+These options will be moved to the `CREATE TABLE .. FROM SOURCE` statements such that
+they can be independently set on a per-table basis. The `envelope` and `encoding` fields
+will be moved from the primary source config to each `SourceExport`, and the source
+rendering pipeline will use per-export encoding/envelope settings rather than assuming
+each export for a source requires the same config, as it currently does.
+
+### Webhook Sources
+
+Webhook sources do not currently behave like other source types, they do not actually
+exist in the storage layer and are not rendered as any sort of dataflow. Instead
+`environmentd` handles incoming HTTP requests bound for webhooks, validates the request
+against any webhook validation options (such as checking an HMAC), and then decodes
+the request body and writes the results directly to a persist collection.
+
+Webhook sources will be migrated to the new model at the SQL level, to allow for
+'schema changes' and multiple outputs using the same `CREATE TABLE .. FROM SOURCE`
+statement syntax as other sources.
+Since the `encoding` and `envelope` options will be moved to that table statement
+this will also allow a webhook request to be decoded differently for each table.
+
+Under the hood, we will continue to have `environmentd` receive Webhook requests
+and write directly to a persist collection. However this code will be modified
+to just do request validation and then write the _raw_ request body as bytes
+and a map of all headers to the persist collection without doing any decoding.
+
+We will then create a new `Persist Source` that operates exactly like all other source
+types such that the storage layer renders a dataflow and the source operator reads
+from an existing persist collection -- in this case the 'raw' webhook collection being
+written to by `environmentd`. This `Persist Source` can have multiple `SourceExports`
+like other sources, and each export can define its own `encoding` and `envelope` processing
+to convert the raw bytes of each request into the appropriate relation schema.
+The progress collection for the webhook source will essentially be the frontier of
+the underlying 'raw' persist collection that each request is written to.
```

Comment:
I think I understand the commonality we're trying to extract from webhook sources here, but I worry it won't be that valuable in practice. It's hard for me to imagine a real world use case for changing the body format, for example. 

Retroactively adding new headers seems more valuable, but I'm worried it's actually quite surprising if we store the full raw HTTP requests for all time, even though that raw collection isn't visible to users. For example we document that you can exclude headers to omit sensitive information: https://materialize.com/docs/sql/create-source/webhook/#excluding-header-fields

But with the proposed design, that sensitive information will sit secretly inside Materialize until the source is dropped—unless someone reveals it with an `ALTER SOURCE` command.

The migration here for existing webhook sources seems inviable, too? Since we don't have the raw HTTP bytes saved for existing webhook sources.

Is there something lighter weight we could do here instead? Something that moves the ball forward without fully aligning with the new subsources model while also not painting ourselves into a corner?  I think the most pressing concern is separating out the webhook source object from the webhook table object. But I think you could reasonably say that as a special case 1) envelopes and encodings are a property of the webhook source, not the subsource and 2) webhook subsources cannot be dropped/added.

Backing up a level, I think what folks actually want with webhook sources is the ability to "write at" webhook tables. Much more so than with other sources. Usually with webhooks you're totally happy to tear down the existing webhook source and make a new one with the new configuration (new headers, new `CHECK`, new body format, whatever) ... if only you could backfill the new webhook source with the fixed up data from the old webhook source!

```
CREATE SOURCE FROM WEBHOOK og_webhook ... HEADER 'foo';

-- Shoot, I need another header going forward!

CREATE SOURCE FROM WEBHOOK webhook2 ... HEADER foo, HEADER 'bar';

-- One time backfill

-- INSERT INTO webhook2 SELECT body, foo, 'unknown' AS bar FROM og_webhook;
```

Reply:
I'm short on time, so that might not be entirely coherent. Happy to chat more about this live next time we're both in the office, @rjobanp, or during this week's S&S sync.

Reply:
Those are all valid concerns and I am definitely worried about the migration of existing webhook sources to the model proposed here.
While we could do something simpler, I worry that we will continue to have special-casing for webhook sources and will need additional special casing for `CREATE TABLE .. FROM SOURCE` statements that refer to webhook sources unless we do the work now to align them to the new model. This isn't just confusing for us in the codebase, but also for users and our documentation.

Following up to our live discussion -- if the main concern is that there might be sensitive data in the raw request bytes we write to the first persist collection, I wonder if we instead can just have the 'Persist Source' ask persist to truncate the 'raw' collection as it ingests data into the downstream 'tables', such that we remove data just after it's been decoded. This would still operate with the new model but would just mean that any new 'table' added to an existing webhook source wouldn't be able to re-ingest previous requests, but all future requests would get decoded individually by all tables from that point forward.